docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,123 @@
+
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `isinst` &ndash; Will always place `null` on stack.
+- `castclass` &ndash; Will always throw `System.InvalidCastException`.
+
+## API Proposal
+
+Support for the following will be indicated by a new property. For .NET 7, the feature will be marked with `RequiresPreviewFeaturesAttribute` to indicate it is in [preview](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md).
+
+```diff
+namespace System.Runtime.CompilerServices
+{
+    public static partial class RuntimeFeature
+    {
++        /// <summary>
++        /// Represents a runtime feature where byref-like types can be used in Generic parameters.
++        /// </summary>
++        public const string GenericsAcceptByRefLike = nameof(GenericsAcceptByRefLike);
+    }
+}
+```
+
+The compiler will need an indication for existing troublesome APIs where ByRefLike types will be permissable, but where the failure will be handled at runtime. An attribute will be created and added to these APIs.
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Indicates to the compiler that constrain checks should be suppressed
+    /// and will instead be enforced at run-time.
+    /// </summary>
+    public sealed class SuppressConstraintChecksAttribute : Attribute
+    { }
+}
+```
+
+Troublesome APIs:
+- [`Span<T>`](https://docs.microsoft.com/dotnet/api/system.span-1)
+    - `public Span(T[]? array);`
+    - `public Span(T[]? array, int start, int length);`
+    - `public T[] ToArray();`
+    - `public static implicit operator Span<T>(ArraySegment<T> segment);`
+    - `public static implicit operator Span<T>(T[]? array);`
+- [`ReadOnlySpan<T>`](https://docs.microsoft.com/dotnet/api/system.readonlyspan-1)
+    - `public ReadOnlySpan(T[]? array);`
+    - `public ReadOnlySpan(T[]? array, int start, int length);`
+    - `public T[] ToArray();`
+    - `public static implicit operator ReadOnlySpan<T>(ArraySegment<T> segment);`
+    - `public static implicit operator ReadOnlySpan<T>(T[]? array);`
+
+A new `GenericParameterAttributes` value will be defined which also represents metadata defined in the `CorGenericParamAttr` enumeration. Space is provided between the existing constraints group to permit constraint growth while keeping the bit-mask contiguous.
+
+```diff
+namespace System.Reflection
+{
+    [Flags]
+    public enum GenericParameterAttributes
+    {
++        AcceptByRefLike = 0x0100
+    }
+}
+```
+
+```diff
+typedef enum CorGenericParamAttr
+{
++   gpAcceptByRefLike = 0x0100 // type argument can be ByRefLike
+} CorGenericParamAttr;
+```
+
```

Comment:
Consider an API currently defined as 

```c#
public void M<T>() { }
```

Given this design consider if it is changed to the following:

```c#
public void M<T>() allow T : ref struct { }
```

Does that represent a binary breaking change or not (assuming C# compiler emits the metadata as specified here)? Think it's overall desirable that this not be a binary breaking change as it allows existing APIs, that could be `ref struct` compatible, to be evolved to be such.

Reply:
I was intending this to follow the same rules as if someone was relaxing an existing constraint. For example, if one converts

```csharp
public void M<T>() where T : unmanaged { }
```
to
```csharp
public void M<T>() { }
```

In my mind this is the same as being more permissible with now allowing `ref struct`. I'd argue the relaxing of constraints and expansion of permissible types should follow the same rules.

I've not checked if the current constraint behavior is a breaking change so perhaps someone can education me here, but that is what I would expect.